### ISSUE TITLE
perf: investigate gRPC packeting optimizations

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#17370]
gRPC creates unnecessary packets during unary response processing. An upstream PR (https://github.com/grpc/grpc-go/pull/1373) to fix this was buggy and rejected and we're waiting on a real fix from the gRPC folks. But above and beyond reducing the number of packets per response to 1, we can look at reducing it below 1 by combining multiple responses into a single packet. Additionally, flushing the write buffer to the connection currently blocks other goroutines. We should investigate a pipelined approach where the write buffer is filled and when it needs to be flushed a new buffer is swapped into place while the buffer is written to the connection.

**Comment by irfansharif** [comment]
could you add more explicit details here? Not as yet familiar with the code paths where this would have a measurable effect or even how I'd go about measuring this? Exec latencies, RTT latencies? 

I also recall some prior experiments you ran investigating this effectively reducing {99,95}-th percentile latencies (?), if there's anything relevant pertaining to this it'd be nice to add.

**Comment by petermattis** [comment]
See https://github.com/grpc/grpc-go/blob/master/transport/http2_server.go#L816 and https://github.com/grpc/grpc-go/blob/master/transport/http2_client.go#L681. The measurable effect of reducing the number of packets (or perhaps the benefit is from reducing the number of system calls) is measurably lower latencies which directly impact performance. Also see https://github.com/grpc/grpc-go/pull/1373 which demonstrated some of the benefits, but had a deadlock scenario. What I'd like to explore here is even more aggressively reducing flushes when there are multiple goroutines sending concurrently. The tricky part here is how to do this without introducing deadlock scenarios caused by the application level flow control. But perhaps experimental code doesn't need to be perfect in order to show whether further investment is worthwhile.

**Comment by petermattis** [comment]
In addition to the Cockroach measured network latencies, I also look at the packet/sec numbers that the prometheus node exporter provides. For a given queries/sec, lower packet/sec is usually better.

**Comment by irfansharif** [comment]
Simply aggregating flushes for a given request into one.
```
___ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)
  +5.81%   +5.67%  -34.65%   -3.98%   -5.44%
```

Before:
![image](https://user-images.githubusercontent.com/10536690/29427790-f0f6e612-8358-11e7-96db-38f55d5aa003.png)

After:
![image](https://user-images.githubusercontent.com/10536690/29427780-e29a202a-8358-11e7-825d-e5003761f75f.png)

Note the reduced number of syscalls, now present on only `WriteStatus{,Optimized}` which takes place for every unary RPC.

**Comment by petermattis** [comment]
This matches what I saw previously. Will be interesting to see what happens if we perform less than 1 syscall per RPC.

**Comment by irfansharif** [comment]
There's a bunch of other low hanging fruit I'm picking through. [Here](https://github.com/grpc/grpc-go/blob/bfaf0423469fc4f95e9eac7e3eec0b2abb46fcca/transport/http2_server.go#L786) `Proto()` does a deep (defensive) copy when it doesn't really need to (Marshalling out to a `io.Writer` doesn't/shouldn't mutate the original proto struct). This deep copy took about as much time as the batched flush itself.

```
         .          .    876:
         .      910ms    877:   if p := st.Proto(); p != nil && len(p.Details) > 0 {
         .          .    878:           stBytes, err := proto.Marshal(p)
         .          .    879:           if err != nil {
         .          .    880:                   // TODO: return error instead, when callers are able to handle it.
         .          .    881:                   panic(err)
         .          .    882:           }
         .          .    883:
         .          .    884:           t.hEnc.WriteField(hpack.HeaderField{Name: "grpc-status-details-bin", Value: encodeBinHeader(stBytes)})
         .          .    885:   }
         .          .    886:
```

**Comment by irfansharif** [comment]
As an aside, gRPC makes for another example where increasing Go's minimum stack size helps considerably:
 ![image](https://user-images.githubusercontent.com/10536690/29430958-439fac40-8364-11e7-8434-895b7521129b.png)

**Comment by irfansharif** [comment]
> beyond reducing the number of packets per response to 1, we can look at reducing it below 1 by combining multiple responses into a single packet

It seems that this is already the case-ish. It's done so by the use of `framer.{adjust}numWriters` and flushing explicitly when you're the last writer-ish. See [here](https://github.com/grpc/grpc-go/blob/596a6acc87c73c3bc398a60ef0089976491ca060/transport/http2_server.go#L918-L931). Unless I'm grossly misunderstanding something, although unlikely in practice, there's a potential starvation issue right there. You can have writers keep coming in with the `adjustNumWriters({0,-1}) == {1,0}` conditionals never becoming true and therefore nothing ever getting flushed. 
Still confused.

**Comment by petermattis** [comment]
When I looked, the `adjustNumWriters` stuff was intended to avoid some flushes, but grpc still always flushed in `WriteStatus` and after sending the body of a unary response. I think `adjustNumWriters` is at the wrong level and it would need to be called before the headers for a unary response are sent and after the status is written. Note that the underlying bufio writer has a limited size. Ideally, we never flush manually and let the bufio flush when it reaches capacity.

PS I see a lot of similarities between flushing the network connection and syncing the WAL.

**Comment by irfansharif** [comment]
- https://github.com/grpc/grpc-go/pull/1478
- https://github.com/grpc/grpc-go/pull/1482
- https://github.com/grpc/grpc-go/pull/1480
- https://github.com/golang/protobuf/pull/418

**Comment by irfansharif** [comment]
heh, https://github.com/grpc/grpc-go/pull/1498 goes some way towards having a dedicated a dedicated flushing goroutine.

**Comment by petermattis** [comment]
@irfansharif Yeah, I noticed that too.

**Comment by petermattis** [comment]
TL;DR? gRPC has very bad scalability with concurrent RPCs. We need to figure out what is going on and either fix it or replace it. Cc @spencerkimball, @bdarnell, @tschottdorf, @a-robinson, @nvanbenschoten 

I extended the `pinger` test program which can be used to send simple "ping" requests and responses over either grpc or another protocol. The following shows using grpc with 1000 concurrent client workers sending 200 byte requests and receiving 200 byte responses as fast as possible (note these tests are running on linux VMs on Azure):

```
~ ./pinger -c 200 -p 1000 -n 1 -d 10s -t grpc cockroach-denim-0001:50051
...
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  40836.9     15.6     28.3     35.7     41.9     75.5
```

Here is the equivalent workload with the "x" (eXperimental) protocol:

```
~ ./pinger -c 200 -p 1000 -n 1 -d 10s -t x cockroach-denim-0001:50051
...
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 335770.9    129.0      6.0      5.8      8.4     18.9
```

I don't have an adequate explanation for the performance difference. "x" is much simpler than grpc and I took some pains to avoid unnecessary flushes of network data. But even accounting for that and trying to mimic the grpc behavior, "x" is still much faster.

Note that 1000 concurrent client workers is extreme and the performance discrepancy doesn't scale down linearly. The following shows grpc scaling from 1 to 512 clients (in powers of 2):

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s    869.7      0.3      0.0      1.5      2.2      8.1
     10s   1838.5      0.7      0.0      1.6      2.2      7.6
     10s   4006.3      1.5      1.0      1.5      2.0     10.0
     10s   7969.3      3.0      0.8      1.6      2.2      8.9
     10s  15422.5      5.9      0.7      1.8      2.6     10.5
     10s  25620.2      9.8      2.1      2.4      3.4     12.1
     10s  32319.0     12.3      2.5      3.7      5.5     15.2
     10s  38490.1     14.7      2.5      6.3      9.4     28.3
     10s  49670.8     18.9      5.0     10.0     13.6     27.3
     10s  45189.6     17.2     12.1     18.9     25.2     44.0
```

And here is the same scaling with "x":

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s    953.4      0.4      0.0      1.4      2.0     10.5
     10s   2060.3      0.8      0.0      1.6      2.5     11.0
     10s   3959.5      1.5      0.9      1.6      2.4     10.5
     10s   8082.6      3.1      0.0      1.6      2.4      8.9
     10s  18925.1      7.3      0.7      1.3      2.2     10.5
     10s  42534.8     16.3      0.7      1.2      2.0      7.1
     10s  79104.9     30.4      0.8      1.3      2.1     11.0
     10s 138325.6     53.2      0.9      1.6      2.6     11.5
     10s 213606.0     82.1      1.0      2.1      3.5     50.3
     10s 279076.3    107.3      1.8      3.4      5.2    113.2
```

NB: I used 200 byte requests and responses because that is approximately the size of a `ScanRequest` and `ScanResponse` for `kv` read operations. Looking at `top` during these runs, I noticed that "x" used ~1/2 the cpu of grpc.

**Comment by tschottdorf** [comment]
Are you testing that against tip? I haven't checked our vendored grpc version, but I thought they had done some performance work (also doubt it'd catch up to X, though).

Are you using single connections for both protocols?

Is your [pinger repo](https://github.com/petermattis/pinger) up to date?

**Comment by petermattis** [comment]
> Are you testing that against tip? I haven't checked our vendored grpc version, but I thought they had done some performance work (also doubt it'd catch up to X, though).

Yes, I'm testing against gRPC tip. The recent performance work provides a very small performance improvement.

> Are you using single connections for both protocols?

Single connections for both protocols. There is a flag to allow using multiple connections, but gRPC doesn't show any benefit from doing that. Before doing this work on `pinger`, I whacked something into CockroachDB to allow using multiple gRPC connections and that also didn't show any benefit.

> Is your pinger repo up to date?

It is now (just pushed).

**Comment by petermattis** [comment]
Hmm, a small tweak to how "x" performs synchronization to more closely mimic grpc causes its performance to drop to almost exactly the same level as grpc...but only on my laptop. Testing between 2 linux boxes over a real network still shows "x" to significantly outperform grpc. 

The synchronization difference is in how grpc and "x" notify the "write loop". In grpc (on tip), this looks like:

```
func loopyWriter(ctx context.Context, cbuf *controlBuffer, handler func(item) error) {
	for {
		select {
		case i := <-cbuf.get():
			cbuf.load()
			if err := handler(i); err != nil {
				return
			}
		case <-ctx.Done():
			return
		}
	hasData:
		for {
			select {
			case i := <-cbuf.get():
				cbuf.load()
				if err := handler(i); err != nil {
					return
				}
			case <-ctx.Done():
				return
			default:
				if err := handler(&flushIO{}); err != nil {
					return
				}
				break hasData
			}
		}
	}
}
```

Essentially, grpc is using channels for synchronization. `cbuf` is an unbounded channel of items. In comparison, "x" uses a mutex, condition variable and a slice. Pseudo-code looks like:

```
func (x *xServerConn) writeLoop() {
	for {
		s.Lock()
		for len(s.pending) == 0 {
		 	s.cond.Wait()
		}
		pending := s.pending
		s.pending = nil
		s.Unlock()

		for _, p := range pending {
		 	if err := handler(p); err != nil {
		 		log.Fatal(err)
		 	}
		}

		if err := handler(&flushIO{}); err != nil {
			log.Fatal(err)
		}
	}
}
```

Switching "x" to use `controlBuffer` resulted in a dramatic slowdown on my laptop, but only on my laptop.

**Comment by petermattis** [comment]
Update, I've been able to slow down "x" between 2 linux machines through the combination of the switch to using a grpc-style `controlBuffer` and `quotaPool`:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  71760.4     27.6     15.7     22.0     27.3     44.0
```

This is still faster than grpc, but I'm likely just missing some additional quota stuff (grpc has transport and stream level quota as well as quota on the number of streams).

For reference, this is the performance I was seeing earlier with the cond-var based synchronization and no quota pool:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 335770.9    129.0      6.0      5.8      8.4     18.9
```

**Comment by tschottdorf** [comment]
The quota pool corresponds to (stream and/or connection-level) window sizes, right? I saw that you have both at 65k initially. Does anything change if you up that to a lot more (especially the connection based one) or is the culprit just the internal overhead of having the quota pools in the first place?

**Comment by petermattis** [comment]
I'm not sure what the culprit is. We're never getting close to the stream quota as we're sending 200 byte requests. We do seem to be hitting the connection quota at high concurrency, but alleviating the limit doesn't provide a significant performance boost:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  52903.9     20.2     16.3     28.3     35.7     58.7
```

I also tried disabling all of the quota code in grpc by commenting it out:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  58959.2     22.5     16.8     25.2     32.5     56.6
```

So clearly something else is still going on inside grpc that is limiting its performance.

**Comment by tschottdorf** [comment]
CPU profile isn't useful?

**Comment by petermattis** [comment]
Heh, I'm looping through all my tools. CPU profile didn't show anything interesting earlier, I'll look again.

**Comment by tschottdorf** [comment]
If that doesn't help, you could consider (and you probably have) strategic commenting out of parts of the quota handling code 🥇

**Comment by petermattis** [comment]
> If that doesn't help, you could consider (and you probably have) strategic commenting out of parts of the quota handling code 🥇

You might have missed above, but I already did exactly that.

**Comment by tschottdorf** [comment]
Oh, (completely) misread that as "I also set the window size to basically infinity".

**Comment by petermattis** [comment]
Per a suggestion from @bdarnell, I checked out whether the usage of http2 as the low-level framing protocol was a cause of the performance difference. To do so, I added a new "y" protocol that uses http2 framing. For reference, here is what we see with "x":

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 335770.9    129.0      6.0      5.8      8.4     18.9
```

And here is baseline grpc:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  45883.1     17.5     23.1     31.5     37.7     65.0
```

And here is "y":

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 280462.6    107.8      3.9      7.3      9.4     18.9
```

I tried to mimic grpc's usage of http2 which means I send both a headers frame and a data frame for every request and response. In http2, headers are compressed using HPACK, so despite sending 7 headers (key/value pairs), the amount of data sent is usually only 7 bytes (plus the http2 framing overhead).

Turns out my usage of http2 was slightly different because I was neglecting to decode the headers. While encoding of the headers is quite cheap, decoding seems to be expensive. I haven't investigated why yet:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 189190.7     72.7      4.5      8.9     11.5     23.1
```

For completeness, this is what we see if we only send data frames:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 349388.8    134.3      3.8      5.8      7.6     23.1
```

This last number is interesting because it is actually faster than the "x" protocol and this is essentially what is sent on the wire by grpc for streaming rpcs. All of the numbers so far have been for using unary RPCs, but this suggests there might be a significant win from using a streaming RPC for `Batch{Request,Response}`.

**Comment by petermattis** [comment]
Here is grpc using a long-lived stream per worker:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  86437.7     33.0     11.5     19.9     25.2     54.5
```

This is approximately twice as fast as unary RPCs with lower latencies. 

If I comment out the grpc connection/stream quota code we get another doubling of performance:

```
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 159498.1     60.8     10.5     14.7     19.9     44.0
```

This is still half the speed of "y" with both lower throughput and higher latencies.

**Comment by bdarnell** [comment]
Using a long-lived stream per worker would help performance but it would have some downsides. Large requests/responses would also block the entire stream while they're being sent, while using a separate stream for each allows them to be multiplexed. It would also prevent (or complicate) efforts to use streaming rpcs to send responses to scan requests.

**Comment by petermattis** [comment]
I was imagining that we'd have a cache of streams per connection. When sending a batch request, we'd get an idle stream or open a new one and then return the stream once done.

**Comment by petermattis** [comment]
Btw, my next step is to run some quick experiments to validate that improving the RPC system here has an effect on an actual cockroach cluster.

**Comment by petermattis** [comment]
I managed to whack the "y" protocol into CockroachDB for use by the `Batch` service. On a local-single node cluster it shows a very nice speedup on a 95% read workload if the local-server optimization is disabled:

```
before:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   60.0s        0         659152        10985.7      1.5      1.2      3.1      6.0     79.7
after:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   60.0s        0         991233        16520.3      1.0      0.8      2.4      4.7     54.5
```

That's a 50% improvement in throughput with smaller latencies.

On a 6-node cluster, the improvement is less dramatic:

```
before:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.1s        0       14488012        48285.1      8.0      4.2     31.5     60.8   4563.4
after:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0       16049741        53490.9      7.2      4.7     27.3     56.6   1140.9
```

This is only a 10.5% improvement in throughput. Not nothing, but less dramatic than I was expecting.

**Comment by petermattis** [comment]
Removing a small bit of unnecessary data copying from "y" results in:

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0       16439583        54792.4      7.0      4.7     25.2     52.4   1040.2
```

**Comment by petermattis** [comment]
Cc @a-robinson

**Comment by dfawley** [comment]
Greetings, grpc-go TL here.  I just heard about this bug, and wanted to chime in.  The grpc-go team is still very actively working on performance improvements.  Our first area of focus a couple months ago was on flow control, which we had serious issues with, but should now be a non-issue and should almost never require manual setting of the initial window sizes.  See [our blog post](https://grpc.io/2017/08/22/grpc-go-perf-improvements.html) for more info on the improvements made here.  If you encounter any problems related to flow control, please [file an issue in our repo](https://github.com/grpc/grpc-go/issues/new).

The "upcoming work" section of the blog is still mostly in progress -- we have batching of writes across streams working now, but with that done, we identified some problems with concurrency caused by the flow control quota and control buffer mechanisms that @petermattis also [called out above](https://github.com/cockroachdb/cockroach/issues/17370#issuecomment-334267707).  We're continuing to iterate on these, and expect to have some improvements in the next month or two.  We are also working to improve memory re-use to reduce garbage collection.

Some settings that may be worth tweaking for now are the IO buffer sizes:
Client:
https://godoc.org/google.golang.org/grpc#WithReadBufferSize
https://godoc.org/google.golang.org/grpc#WithWriteBufferSize
Server:
https://godoc.org/google.golang.org/grpc#ReadBufferSize
https://godoc.org/google.golang.org/grpc#WriteBufferSize

These buffers have historically been 32KB, which is only 2 frames, and we've found increasing this to 128KB or 256KB can make a big difference in reducing syscalls.  We [just added](https://github.com/grpc/grpc-go/pull/1544) these as options a couple weeks ago, and we intend to change the default to work better out-of-the-box in the near future.

Thanks for digging into these performance issues and even identifying some root causes.  I'm sorry the current state of our performance isn't meeting your expectations, but this is our main area of focus at the moment, so we hope to have some solid improvements soon.  Please feel free to reach out to us (or cc me on your bugs) if you encounter anything that looks wrong, or if you have any questions.

**Comment by petermattis** [comment]
Thanks for the updates @dfawley. I've been tracking the work you guys are doing. While I suspect increasing the read and write buffer size won't help the scenarios I've been testing, I will give them a whirl.

Two changes you might want to consider:
1. Somehow avoid doing anything with the stream quota pool if the size of a unary request or response is less than stream quota size. Simply selecting from the channel seems to cause a slowdown. The other change is to optimize decoding of http2 headers. 
2. In my testing, encoding the headers is fast, but decoding them causes a slowdown. Since the encoded headers are small, I think there could be a fast-path for decoding them as a chunk instead of field-by-field.

**Comment by dfawley** [comment]
> In my testing, encoding the headers is fast, but decoding them causes a slowdown

I overlooked this point in my initial read-through of this issue.  We'll check it out -- the results you showed above are very significant, and seems surprising.

**Comment by a-robinson** [comment]
Hi @dfawley, how has the grpc-go performance work been going? Has there been any progress on the improvements you mentioned were on the way or on understanding areas for improvement?

On a related note, we've been particularly surprised to see just how [much faster the C++ gRPC implementation is in comparison to grpc-go](https://performance-dot-grpc-testing.appspot.com/explore?dashboard=5636470266134528). Is the reason for the 3x-4x difference in throughput understood? It seems as though there are probably lessons that could be carried over to help improve grpc-go.

**Comment by wormwang** [comment]
Are you test brpc https://github.com/brpc/brpc
The developer of brpc show in a benchmark, the brpc is best perf of all rpc.
And, brpc support OpenUCX in plan. OpenUCX support RDMA

**Comment by a-robinson** [comment]
Hi @wormwang, I don't believe we've tested it, but the cgo interface makes using any C++ RPC implementation from go fairly impractical. I did a test where I tried using the C++ implementation of grpc via cgo but its throughput capped out quite low, most likely due to the problems all the blocked cgo threads caused the go scheduler.

**Comment by a-robinson** [comment]
Any updates, @dfawley?

**Comment by dfawley** [comment]
Yes, I have some info I can share.  We made some optimizations in the write path to make it significantly faster, but Go scheduling behavior actually made the resulting overall RPC performance _worse_.  We are fairly confident the issue is that all streams are blocking on the same channel to do flow control checking, which results in those goroutines being scheduled in the same thread, and makes batching of writes less effective.  With that in mind, we are moving the flow control checks into the goroutine that does the framing and network write instead of distributing it across the streams.  We believe this will alleviate the scheduling problem, and should show some good performance gains.  We also intend to make writing to the control buffer lock-free, so if the previous change isn't enough to fix the scheduling/batching problems, this definitely should.

**Comment by a-robinson** [comment]
Good to hear! Thanks for the update, @dfawley. Do you have a rough estimate of when these changes will start materializing?

I'd also like to offer our help if you want it, both for the engineering work of making the improvements or for testing out any unfinished changes you'd like more mileage on. This is quite important to us, so we're happy to devote time to it.

**Comment by MakMukhi** [comment]
@a-robinson There has actually been some improvement since we last updated this issue. This [PR](https://github.com/grpc/grpc-go/pull/1636) showed a significant bump in both our as well as the benchmark mentioned here.
Also, currently we are working on a scheduler for gRPC-Go transport to essentially make writing lock-free. We are quite hopeful with that and should have some concrete performance numbers soon.

**Comment by a-robinson** [comment]
Wow, there has certainly been some upstream progress, at least on the https://github.com/petermattis/pinger benchmark. These are my results from running between two beefy GCE VMs in the same zone:

```
$ ./pinger -p 100 -n 1 -d 10s

head (cd563b81ec334abe62cda34a33b8640ea67ef5ba):
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 183795.7      3.5      0.9      1.0      1.8      5.8

1.8.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s 177266.9      3.4      0.5      1.0      1.8     11.0

1.7.3:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  86291.4      1.6      2.2      2.2      3.0      9.4

1.6.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  62914.8      1.2      0.0      2.9      3.8     15.2
```

Single-client throughput has gotten slightly worse, mostly due to some change between 1.6 and 1.7, but I'll take that trade any day of the week:

```
$ ./pinger -p 1 -n 1 -d 10s

_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   6883.5      0.1      0.1      0.2      0.3      6.3

1.8.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   7018.4      0.1      0.2      0.2      0.3      3.1

1.7.3:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   6515.3      0.1      0.1      0.2      0.3      3.3

1.6.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   7635.1      0.1      0.1      0.2      0.2      3.1
```

I'll try testing out the affect on real CockroachDB performance next, but kudos on the benchmark improvements!

**Comment by a-robinson** [comment]
Things are much less rosy on a real cluster -- some improvement, some regression, and nowhere near the performance of the simpler "cRPC" protocol from https://github.com/cockroachdb/cockroach/pull/19098. All this testing was done on a 6 node roachprod cluster on GCE (all in the same zone, n1-standard-4 VMs, local SSD):

```
$ ./kv --read-percent=95
grpc-head:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   60.5s        0         458665         7587.0      2.1      0.9     15.7     30.4    226.5

grpc-1.7.3:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   70.4s        0         524855         7458.9      2.1      0.9     16.3     30.4    159.4

crpc:
$ ./kv --read-percent=95
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   60.2s        0         501613         8328.3      1.9      0.7     16.3     30.4    285.2
```

```
$ ./kv --splits=17 --read-percent=95
grpc-head:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   20.6s        0         145975         7076.8      2.3      0.8     15.7     35.7    419.4

grpc-1.7.3:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   65.5s        0         520663         7954.2      2.0      0.8     14.2     31.5    234.9

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   65.5s        0         544615         8313.0      1.9      0.6     15.2     33.6    419.4
```

```
$ roachperf alex-rpc wipe
```

```
$ ./kv --read-percent=95 --concurrency=100
grpc-head:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   60.4s        0         883329        14623.4      6.8      2.9     46.1     92.3   1040.2

grpc-1.7.3:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   68.5s        0         701368        10235.4      9.8      4.2     60.8    134.2    637.5

crpc:
$ ./kv --read-percent=95 --concurrency=100
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   64.6s        0        1459569        22609.0      4.4      1.3     37.7     65.0   1073.7
```

```
$ ./kv --splits=16 --read-percent=95 --concurrency=100
grpc-head:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  157.8s        0        3185502        20184.7      5.0      1.6     23.1     88.1   2281.7

grpc-1.7.3:
$ ./kv --splits=16 --read-percent=95 --concurrency=100
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   63.7s        0        1209611        18985.7      5.3      1.5     24.1    100.7    939.5

crpc:
$ ./kv --splits=16 --read-percent=95 --concurrency=100
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   61.1s        0        1911006        31255.3      3.2      0.9     26.2     52.4   1811.9
```

**Comment by a-robinson** [comment]
The local server optimization was disabled in all of the above tests, as well.

**Comment by dfawley** [comment]
Thanks for providing some real-world numbers.  We are specifically working to improve the high-concurrency performance and hope to have some solid improvements there in the next week or so.

Can you explain what the workload is in the `./kv --splits=17 --read-percent=95` test case, so we can try to determine why there was such a significant performance drop from 1.7.3?  I notice it ran for less time on head than 1.7.3 (20s vs. 65s) -- could that explain the difference, i.e. due to spending a higher percentage of time warming up?

**Comment by a-robinson** [comment]
> Can you explain what the workload is in the ./kv --splits=17 --read-percent=95 test case, so we can try to determine why there was such a significant performance drop from 1.7.3? I notice it ran for less time on head than 1.7.3 (20s vs. 65s) -- could that explain the difference, i.e. due to spending a higher percentage of time warming up?

There isn't much warmup involved, but that's probably the least important of all the test cases, and some randomness in where the data got placed in the cluster could potentially affect it. The `--splits` flag just attempts to ensure that data is better spread around the nodes in the cluster. The first two cases I tested (`./kv --read-percent=95` and `./kv --splits=17 --read-percent=95`) are entirely limited by latency, not throughput, since there are so many fewer clients sending concurrent requests, so it's possible that the the latency under low-medium contention is somewhat worse.

I wouldn't worry too much about that one in particular, though.

**Comment by a-robinson** [comment]
So I realized that the benchmark performance we really care about isn't up as much as I made it sound in my comment on Monday. Unary RPCs are most important for our performance, and I had forgotten that  https://github.com/petermattis/pinger defaults to using streaming RPCs, so the benchmark results I pasted were using streaming RPCs. With unary RPCs the improvement since 1.7.3 has been much less dramatic, which better matches the performance I saw in the full cluster testing:

```
$ ./pinger -p 100 -n 1 -d 10s

head (cd563b81ec334abe62cda34a33b8640ea67ef5ba):
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  51998.3      1.0      1.7      3.3      3.5      7.3

1.8.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  52962.1      1.0      2.5      3.1      3.5      6.0

1.7.3:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  46339.5      0.9      1.8      3.5      4.1      6.3

1.6.0:
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  36498.5      0.7      2.6      4.5      5.5      9.4
```

**Comment by MakMukhi** [comment]
Hey guys! We have been working on this issue quite a bit lately to try and reduce lock contention. We have a change that is in review right now but I wanted to create an external diff(base isn't the head though but it isn't too long in past) to share with you guys. Let us know if it's helpful and if you have any feedback.
We have ran @petermattis 's pinger benchmark on it.

Before:
```
./pinger -p 100 -n 1 -d 10s -t grpc
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  84477.8      1.6      1.2      2.0      2.9      7.1

./pinger -p 100 -n 1 -d 10s -t grpc -c 200
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  79687.5     30.4      3.7      2.2      2.9      7.3

./pinger -p 100 -n 1 -d 10s -t grpc -c 65536
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   5599.6    699.9     25.2     33.6     41.9     83.9
```

After:
```
./pinger -p 100 -n 1 -d 10s -t grpc
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  97888.5      1.9      1.0      1.8      2.5     11.0

./pinger -p 100 -n 1 -d 10s -t grpc -c 200
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s  90911.7     34.7      2.1      2.0      2.8      8.9

./pinger -p 100 -n 1 -d 10s -t grpc -c 65536
_elapsed____ops/s_____MB/s__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
     10s   5758.9    719.9     17.8     23.1     26.2     48.2
```

**Comment by petermattis** [comment]
@MakMukhi Thanks. We're busy burning down bugs for our next release for the next few weeks. Will definitely take a look at this after that.

**Comment by MakMukhi** [comment]
@petermattis Hey just looping back to let you guys know that the aforementioned [changes](https://github.com/grpc/grpc-go/pull/1962) were merged recently. 
When you guys get a chance to try it out, let us know if they are helpful. If not, we'd love some feedback either way.
Also, we are following this with a [read-path optimization](https://github.com/grpc/grpc-go/pull/1987) to reduce memory usage. I'll post some number on that soon.

**Comment by a-robinson** [comment]
Thanks @MakMukhi! I'll give the changes a test in cockroach soon.

**Comment by petermattis** [comment]
@MakMukhi Thanks for the heads up. We're following along with the work you're doing and it is on our TODO list to test out the recent improvements. We'll let you know of any results.

**Comment by a-robinson** [comment]
I ran a few experiments today comparing cockroachdb at its current head (e7d98d09cc5df5f1c2f70a8003076378a5d088c5) using 3 different rpc implementations -- grpc 1.9.2 (the default for us right now), the tip of the grpc repo (https://github.com/grpc/grpc-go/commit/5d8897144f741c0358f5f194503db84142d0a985), and @petermattis's old crpc implementation from https://github.com/cockroachdb/cockroach/pull/19098.

The recent grpc changes appear to have no effect at low concurrency levels and a 5-10% throughput improvement at high concurrency levels. The crpc implementation still provides a very large improvement (10-35%), as in past testing.

Used a 6-node default roachprod cluster (n1-standard-4s with local SSD) running `roachprod test kv_95` from a 7th node for all tests.

```
concurrency=6
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        1461325         4871.1      1.2      0.8      6.6     13.1    134.2

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        1459706         4865.7      1.2      0.8      7.1     13.1    142.6

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        1608805         5362.7      1.1      0.7      6.8     13.1     83.9

concurrency=12
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        2732111         9107.0      1.3      0.8      7.6     14.2    130.0

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        2678651         8928.8      1.3      0.8      7.9     14.2    226.5

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        3075721        10252.4      1.2      0.7      7.6     14.2    176.2

concurrency=18
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        3805526        12685.1      1.4      0.9      8.4     14.7    159.4

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        3796336        12654.4      1.4      0.9      7.9     14.7    142.6

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        4262204        14207.3      1.3      0.7      8.1     14.2    184.5

concurrency=24
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        4772702        15908.9      1.5      0.9      8.9     15.2    176.2

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        4773829        15912.7      1.5      0.9      8.4     15.2    285.2

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        5461114        18203.6      1.3      0.7      8.4     14.7    104.9

concurrency=384
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        8579966        28598.2     13.4      3.8     39.8    268.4   2147.5

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        9170500        30566.3     12.6      5.5     41.9    184.5   1476.4

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0       11069178        36894.2     10.4      3.0     32.5    209.7   1409.3

concurrency=concurrency=768
1.9.2:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.1s        0        8298184        27655.7     27.7      4.7     50.3    738.2   9663.7

grpc-tip:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0        9157505        30520.3     25.2      7.1     52.4    520.1   5637.1

crpc:
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s        0       11210516        37363.6     20.6      4.5     67.1    453.0   2550.1
```

**Comment by petermattis** [comment]
@a-robinson Interesting. Thanks for the updated numbers. The improvement to p99 latencies at higher concurrency is significant.

**Comment by MakMukhi** [comment]
@a-robinson Thanks for running the numbers.
This recent effort was primarily about optimizing gRPC-Go in high-concurrency environment by reducing synchronization cost. Although, in our benchmarks we did see significant improvement in low-concurrency unary RPC scenarios. 
We're glad that you could get 10-15% improvement on your application-level benchmarks.

Regarding cRPC:
Is it something that you guys are actively pursuing and  matching its performance is the end-goal of this issue?

The main requirements mentioned in the issue description have been met; 1. batching of messages, headers and other control frames. 2. Not blocking other writers when writing out on the connection.

We're actively still working on performance, in fact there's a [PR](https://github.com/grpc/grpc-go/pull/1987) out for reducing memory usage. However, our next effort would be to optimize internal use cases. In the meanwhile, If you guys have specific performance related goals, we'd be happy to hear about them and eventually match them, hopefully.

**Comment by a-robinson** [comment]
> Is it something that you guys are actively pursuing and matching its performance is the end-goal of this issue?

We aren't actively pursuing it, and don't have any short-term plans to flesh it out into a truly usable implementation. At this point it primarily serves to provide an estimate of how good performance could be if we were to make our rpc layer as fast as possible, which helps prioritize the importance of improving it.

> The main requirements mentioned in the issue description have been met; 1. batching of messages, headers and other control frames. 2. Not blocking other writers when writing out on the connection.
> 
> We're actively still working on performance, in fact there's a PR out for reducing memory usage. However, our next effort would be to optimize internal use cases. In the meanwhile, If you guys have specific performance related goals, we'd be happy to hear about them and eventually match them, hopefully.

This issue has grown beyond the initial optimizations mentioned to being more of a catch-all for RPC performance. We remain very interested in better gRPC performance, particularly in closing the gap between the performance of streaming RPCs (which has gotten quite good) and unary RPCs (which is still quite underwhelming by comparison).

Improved memory usage should certainly help. What sort of details would be helpful to you? A more realistic benchmarking program?

**Comment by MakMukhi** [comment]
I believe, although cRPC gives a good lower bound on numbers, a comparison between cRPC and gRPC might not be fair. I say that because the very features that cRPC is missing (flow control, context cancellation propagation, chunking data into fixed-size frames, goaway handling etc.) have performance penalties that gRPC incurs.

Yes! If you guys can provide a  realistic application-level benchmark that'd be awesome.

**Comment by petermattis** [comment]
> I believe, although cRPC gives a good lower bound on numbers, a comparison between cRPC and gRPC might not be fair. I say that because the very features that cRPC is missing (flow control, context cancellation propagation, chunking data into fixed-size frames, goaway handling etc.) have performance penalties that gRPC incurs.

Yeah, we're aware that the comparison between cRPC and gRPC isn't really fair, and we're aware that cRPC lacks a lot of features we'd need to switch to it, yet the comparison is informative. I'm still somewhat surprised that the addition of the functionality you mention (and others) has such a significant performance impact. 

> Yes! If you guys can provide a realistic application-level benchmark that'd be awesome.

Ack. Not sure when we'd be able to provide this. An initial step would be for us to gather some metrics and data from the perf test Alex is running that we (or you) could use to power a synthetic workload. For example, we could get stats about the request and response sizes and the distribution of processing times. That seems like it would be sufficient info to fold into an enhanced version of `pinger`.

**Comment by petermattis** [comment]
PS @MakMukhi we're very appreciative of the performance improvements you've been making to gRPC. We're hoping we can put off doing anything on cRPC long enough for you to sufficiently close the performance gap so that we never have to do that work. 😄

**Comment by MakMukhi** [comment]
Yeah that'll be a good start. If we get a good enough picture of the workload/scenario we can update `pinger` accordingly. 
Thanks for the appreciation. I feel very optimistic about for gRPC's future.
